packages/babel-parser/src/tokenizer/index.js

```diff
@@ -990,15 +991,15 @@ export default class Tokenizer extends LocationParser {
       this.state.pos - start >= 2 &&
       this.input.charCodeAt(start) === charCodes.digit0;
     if (octal) {
+      const number = this.input.slice(start, this.state.pos);
       if (this.state.strict) {
         this.raise(
           start,
           "Legacy octal literals are not allowed in strict mode",
         );
       }
-      if (/[89]/.test(this.input.slice(start, this.state.pos))) {
-        octal = false;
-      }
+      if (/[89]/.test(number)) octal = false;
+      if (/^[0-9]*$/.test(number)) isNonOctalDecimal = true;
```

Comment:
`octal == true` implies that `startsWithDot == false` (because it starts with zero), which in turn means that `this.readInt(10) != null`, or otherwise we would throw `Invalid number` error here.
```js
if (!startsWithDot && this.readInt(10) === null) {
      this.raise(start, "Invalid number");
}
```
IIRC `this.readInt(10)` would ensure the input from `start` to `this.state.pos` to always be `[0-9]`, if so we don't have to check `[0-9]` again, which means `isNonOctalDecimal` equals to the previous definition of `octal`:
```js
this.state.pos - start >= 2 && this.input.charCodeAt(start) === charCodes.digit0;
```

Reply:
Yeah, so it would be enought to do
```js
if (/[89]/.test(number)) {
  octal = false;
  isNonOctalDecimal = true;
}
```

So that `isNonOctalDecimal` literally means "we tought that it was an octal number, but it turns out that it isn't"

Reply:
You're right, i didn't see it in that way, it took me a while to realize why the test i added was passing, but it's because when `octal` is set to false here:

```javascript 
if (/[89]/.test(number)) octal = false;
```
`isNonOctalDecimal` remains true, so it can throw the error here:

```javascript
 if (isFloat || octal || isNonOctalDecimal) {...}
```
It's confusing, thanks for the review, i'll come back with some changes!

Reply:
@nicolo-ribaudo Nice! Also, the `decimal` part isn't confusing? Sounds like fractional numbers, maybe it can only be `isNonOctal`?

Reply:
The spec calls it [`NonOctalDecimalIntegerLiteral`](https://tc39.es/ecma262/#prod-annexB-NonOctalDecimalIntegerLiteral), so maybe `nonOctalDecimalInt`?

But if you prefer `isNonOctal`, it's ok :+1: